src/transport/MessageCounter.h

```diff
@@ -88,13 +90,28 @@ class LocalSessionMessageCounter : public MessageCounter
     LocalSessionMessageCounter() { mValue = (Crypto::GetRandU32() & kMessageCounterRandomInitMask) + 1; }
 
     Type GetType() const override { return Session; }
+    bool IsValid() const override { return mValue != kInvalidMessageCounter; }
     uint32_t Value() const override { return mValue; }
     CHIP_ERROR Advance() override
     {
+        if (mValue == kInvalidMessageCounter)
+        {
+            return CHIP_ERROR_MESSAGE_COUNTER_EXHAUSTED;
+        }
+
+        if (mValue == kMaxMessageCounter)
```

Comment:
Yes, `Advancing 0xFFFFFFFF is ok and will get 0`, this is true and expected.

Here is how the counter works:
1. Verify the value is valid
2. Get the value
3. Use the value
4. Advance the value

There are several facts:
1. 0xFFFFFFFF is an valid value
2. The value must be changed to a different value after advance
3. The error returned by `Advance` doesn't matter, because it doesn't affect encoding the current message

This end up with current implementation.


I also considered other implementations, like:
1. Advance the value (return error if can't advance)
2. Get the value
3. Use the value

This is also a valid approach, I can reform it to this if it is preferred. @andy31415

Reply:
I'd prefer the later approach where the Init simply does adds a `-1` step in anticipation of `Advance` being called.  The current logic, while being correct, is counter-intuitive, which is not a great property in security code.